sdk/python/kfp/v2/dsl/component_bridge.py

```diff
@@ -136,15 +139,23 @@ def _resolve_output_path_placeholder(output_key: str) -> str:
     else:
       return _output_artifact_placeholder(output_key)
 
-  # IR placeholders are decided merely based on the declared type of the input.
-  # It doesn't matter wether it's InputValuePlaceholder or InputPathPlaceholder
-  # from component_spec.
-  placeholder_arguments = {
-      input_spec.name: _input_artifact_placeholder(input_spec.name)
-      if type_utils.is_artifact_type(input_spec.type) else
-      _input_parameter_placeholder(input_spec.name)
-      for input_spec in component_spec.inputs or []
-  }
+  placeholder_arguments = {}
+  for input_spec in component_spec.inputs or []:
+    # Skip inputs not presented in arguments list. This could happen for
+    # optional inputs whose presence in command/arguments is controlled by
+    # if placeholder.
+    if input_spec.name not in arguments:
```

Comment:
qq: What if this is true and this input is not controlled by if-placeholder? Shall we raise an error instead?

Reply:
The comment here is misleading, I'll change that.

To answer the question: If the input is required and it's not in this list, then there will be an error. Otherwise, if it's optional, then no matter whether it's guarded by if placeholder, it's allowed to be absent. The difference is that for an absent input, if it's not guarded by if placeholder, then it won't appear in `container_spce.args`/`container_spec.command` anyway. So it has already been filtered out.

Reply:
I see. To confirm my understanding is correct, we only look at what's actually used as placeholders and skip those are not. Right?

Feel free to resolve if my understanding is correct.

Reply:
That's correct.

Reply:
>I see. To confirm my understanding is correct, we only look at what's actually used as placeholders and skip those are not.

I'd say the situation is as follows:
The `_resolve_command_line_and_paths` fails if it needs to resolve an `inputValue` placeholder, but the argument is missing while the input is not optional and does not have a default value.
The code here just passes through the arguments, converting them. This would be more clear if the code was just iterating over the `arguments` (as opposed to iterating over inputs).